src/System.Numerics.Tensors/src/System/Numerics/Tensors/Experimental/IReadOnlyTensor.cs

```diff
@@ -0,0 +1,20 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Numerics.Tensors
+{
+    public interface IReadOnlyTensor<T>
+    {
+        long Rank { get; }
+
+        T this[params long[] indices] { get; }
+        T this[ReadOnlySpan<long> indices] { get; }
+
+        long GetDimensionLength(long dimension);
+        IReadOnlyTensor<T> Slice(params (long start, long length)[] ranges);
```

Comment:
Instead of a Tuple, what do you think about introducing a new `struct` for this?

If we stick with Tuple, I thought the recommendation was for PascalCasing of the members.

Reply:
I'd rather fix the casing of the names here. I don't think introducing a second `TensorRange` type makes sense since it would either need to mirror the existing `Range` and have the methods take `TensorRange[]` or it would need to internally track `long startIndices[]` and `long lengths[]` (which would make the type need to be a `ref struct` if we wanted to make it efficient).

Reply:
My thoughts are:

1. There is no "existing `Range`" type. There is a newly proposed one, but that hasn't shipped and isn't available here.
2. Mirroring the newly proposed `Range` type IS beneficial because it is not confusing to users. When working with System.Range, I deal with "start" and "end".  When working with Tensor, I deal with "start" and "length".
3. When `Range` does become available to Tensor to use someday, there is a path to make `TensorRange` convertible to/from `System.Range` and we have some sort of hope that the C# range feature could maybe work with this API. Using `ValueTuple<long, long>` there is no hope of that ever happening - because who would want `ValueTuple<long, long>` convertible to/from `System.Range`?

---
In reply to: [265283292](https://github.com/dotnet/corefx/pull/36019#discussion_r265283292) [](ancestors = 265283292)

Reply:
> Mirroring the newly proposed Range type IS beneficial because it is not confusing to users. When working with System.Range, I deal with "start" and "end". When working with Tensor, I deal with "start" and "length".

This is something users already have to rationalize. `Range` deals with two `Index` types that are abstract from the underlying type (this is so they can do things like, 5 from End). Most existing APIs, however, deal with `start` and `length` as `integer` types. If you want to support both, you will likely have a method which takes a `Range` and all it does is finalizes the indices to be a `start` and `length` and then you call the existing overload that takes a `start` and a `length`.

Reply:
If we were to leave it as tuples, shouldn't the name of each element be named using PascalCasing? I was naming each element by PascalCasing as it appears like any other type members in an IDE.